algebird-core/src/main/scala/com/twitter/algebird/MomentsGroup.scala

```diff
@@ -102,6 +146,96 @@ sealed class Moments(val m0D: Double, val m1: Double, val m2: Double, val m3: Do
 }
 
 object Moments {
+  final class MomentsState(
+      var count: Double,
+      var mean: Double,
+      var m2: Double,
+      var m3: Double,
+      var m4: Double
+  ) {
+
+    def +=(b: Moments): this.type = {
+      /*
+       * Unfortunately we copy the code from the monoid's plus implementation,
+       * but we do it to avoid allocating a new Moments on every item in the
+       * loop. the Monoid laws test that sum matches looping on plus
+       */
+      val countCombined = count + b.m0D
+
+      if (countCombined == 0.0) {
+        mean = 0.0
+        m2 = 0.0
+        m3 = 0.0
+        m4 = 0.0
+      } else {
+        val delta = b.mean - mean
+        val delta_n = delta / countCombined
+        val delta_n2 = delta_n * delta_n
+        val delta_n3 = delta_n2 * delta_n
+        val count_sq = count * count
+        val rn_sq = b.m0D * b.m0D
+
+        val meanCombined = Moments.getCombinedMeanDouble(count, mean, b.m0D, b.mean)
+
+        val m2Combined = m2 + b.m2 + delta * delta_n * count * b.m0D
+
+        val m3Combined = m3 + b.m3 +
+          delta * delta_n2 * count * b.m0D * (count - b.m0D) +
+          3 * delta_n * (count * b.m2 - b.m0D * m2)
+
+        val m4Combined = m4 + b.m4 +
+          delta * delta_n3 * count * b.m0D *
+          (count_sq - count * b.m0D + rn_sq) +
+          6 * delta_n2 * (count_sq * b.m2 + rn_sq * m2) +
+          4 * delta_n * (count * b.m3 - b.m0D * m3)
+
+        mean = meanCombined
+        m2 = m2Combined
+        m3 = m3Combined
+        m4 = m4Combined
+      }
+
+      count = countCombined
+      this
+    }
+
+    def +=(b: Double): this.type = {
+      val prevCount = count
+      count += 1
+
+      val delta = b - mean
+      val delta_n = delta / count
+      val delta_n2 = delta_n * delta_n
+      val term1 = delta * delta_n * prevCount
+
+      mean = Moments.getCombinedMeanDouble(prevCount, mean, 1.0, b)
+      m4 += term1 * delta_n2 * (count * count - 3 * count + 3) +
+        6 * delta_n2 * m2 - 4 * delta_n * m3
+      m3 += term1 * delta_n * (count - 2) - 3 * delta_n * m2
+      m2 += term1
+      this
+    }
+
+    def toMoments: Moments = new Moments(count, mean, m2, m3, m4)
+
+    def resetFromMoments(m: Moments): this.type = {
+      count = m.m0D
+      mean = m.m1
+      m2 = m.m2
+      m3 = m.m3
+      m4 = m.m4
+      this
+    }
+  }
+
+  object MomentsState {
+    def fromMoments(m: Moments): MomentsState =
+      new MomentsState(m.m0D, m.m1, m.m2, m.m3, m.m4)
+
+    def newEmpty: MomentsState =
```

Comment:
nit: scala has developed a style of using `def newEmpty(): MomentsState` for non-pure things (in this case we are allocating mutable memory, so it is not pure, and the `newEmpty()` helps signal that.